File: 프로그래머스/lv3/이재표/다단계 칫솔 판매.java
import java.util.*;
class Solution {
    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        int n = enroll.length;
        int[] answer = new int[n];
        Map<String, Integer> map = new HashMap<>();

        for (int i = 0; i < n; i++) {
            map.put(enroll[i], i);
        }

        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            if (referral[i].equals("-")) {
                arr[i] = -1;
            } else {
                arr[i] = map.get(referral[i]);
            }
        }

        for (int i = 0; i < seller.length; i++) {
            int sellerIndex = map.get(seller[i]);
            int p = sellerIndex;
            int totalAmount = amount[i] * 100;

            while (p != -1) {
                int up = totalAmount / 10;
                answer[p] += totalAmount - up;
                totalAmount = up;
                p = arr[p];
                if (totalAmount == 0) {break};
            }
        }

        return answer;
    }

}

File: lv3/다단계 칫솔 판매.py
def dfs(name, money, total, graph):
    global left_money
    left_money = int(money * 0.1)  # 10% 추천인
    total[name] += (money - left_money)  # 90% 가지기

    # 종료 조건 center 하위 노드 or 1원 미만
    if graph[name] == "-" or left_money == 0:
        return
    # 추천인에게 올라가기
    else:
        dfs(graph[name], left_money, total, graph)


def solution(enroll, referral, seller, amount):
    total, graph, answer = {}, {}, []

    for i in range(len(enroll)):
        graph[enroll[i]] = referral[i]
        total[enroll[i]] = 0

    left_money = 0

    for i in range(len(seller)):
        dfs(seller[i], amount[i] * 100, total, graph)
    for i in enroll:
        answer.append(total[i])

    return answer